### PR TITLE
Added OCPP 2.0.1 Calculate Composite Schedule functionality and tests.

### PR DESCRIPTION
**Pull request by folkengine**
## Describe your changes

Adding functionality for calculating composite schedule into OCPP 2.0.1 as well as unit tests
for OCPP 1.6 calculate enhanced composite schedule

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [ ] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: include/ocpp/v201/smart_charging.hpp [hunk 1]
```diff
@@ -85,6 +106,40 @@ class SmartChargingHandler {
...
+    ///
+    /// \brief Iterates over the periods of the given \p valid_profiles and determines the earliest next absolute period
+    /// end time later than \p temp_time
+    ///
+    ocpp::DateTime get_next_temp_time(const ocpp::DateTime temp_time,
+                                      const std::vector<ChargingProfile>& valid_profiles, const int32_t evse_id);
+
+    ocpp::DateTime get_period_end_time(const int period_index, const ocpp::DateTime& period_start_time,
+                                       const ChargingSchedule& schedule);
+
```
Comment (ghost) [inline include/ocpp/v201/smart_charging.hpp]: Based on previous feedback anything that is not `calculate_composite_schedule` EVerest will want to have it `protected` if it is not being directly used.

File: lib/ocpp/v201/smart_charging.cpp [hunk 2]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
 }
 
+void log_period_date_time_pair(PeriodDateTimePair period_date_time_pair) {
+    std::string log_str = "PeriodDateTimePair> ";
+
+    if (period_date_time_pair.period.has_value()) {
+        log_str += " period: " + utils::to_string(period_date_time_pair.period.value());
+        // log_str += " period: " + std::to_string(period_date_time_pair.period.value().startPeriod);
+        // EVLOG_info << "   Period: " << period_date_time_pair.period.value().startPeriod
+        //            << " End Time: " << period_date_time_pair.end_time.to_rfc3339();
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: NIT: Can this comment be removed?

File: lib/ocpp/v201/smart_charging.cpp [hunk 3]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+}
+
+CompositeSchedule
+SmartChargingHandler::initialize_enhanced_composite_schedule(const ocpp::DateTime& start_time,
+                                                             const ocpp::DateTime& end_time, const int32_t evse_id,
+                                                             ChargingRateUnitEnum charging_rate_unit) {
+    CompositeSchedule composite_schedule;
+
+    composite_schedule.evseId = evse_id;
+    composite_schedule.duration = duration_cast<seconds>(end_time.to_time_point() - start_time.to_time_point()).count();
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: Should this be `determine_duration`?

File: lib/ocpp/v201/smart_charging.cpp [hunk 4]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    // EVLOG_info << "";
+    // EVLOG_info << "time_window> " << temp_time.to_rfc3339() << " - " << end_time.to_rfc3339() << " = " << count << "
+    // ("
+    //            << get_log_duration_string(count) << ")";
+    // if (count > 0) {
+    //     return true;
+    // } else {
+    //     EVLOG_info << "time_window> count = " << count << " RETURN FALSE ";
+    //     return false;
+    // }
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: NIT: Can this comment be removed?

File: include/ocpp/v201/smart_charging.hpp [hunk 5]
```diff
@@ -125,6 +180,17 @@ class SmartChargingHandler {
     std::vector<ChargingProfile> get_station_wide_tx_default_profiles() const;
     void conform_validity_periods(ChargingProfile& profile) const;
     CurrentPhaseType get_current_phase_type(const std::optional<EvseInterface*> evse_opt) const;
+    CompositeSchedule initialize_enhanced_composite_schedule(const ocpp::DateTime& start_time,
+                                                             const ocpp::DateTime& end_time, const int32_t evse_id,
+                                                             ChargingRateUnitEnum charging_rate_unit);
```
Comment (christopher-davis-afs) [inline include/ocpp/v201/smart_charging.hpp]: What makes this "enhanced"?

File: lib/ocpp/v201/smart_charging.cpp [hunk 6]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+        time_point<date::utc_clock> period_end_time;
+        for (size_t i = 0; i < periods.size(); i++) {
+            const auto period_end_time = get_period_end_time(i, period_start_time.value(), schedule);
+
+            EVLOG_info << "   find_period_at>        start_time> " << time.to_rfc3339();
+            EVLOG_info << "   find_period_at> period_start_time> " << period_start_time.value().to_rfc3339();
+            EVLOG_info << "   find_period_at>   period_end_time> " << period_end_time.to_rfc3339();
+
+            if (time >= period_start_time.value() && time < period_end_time) {
+                PeriodDateTimePair date_time_pair = {periods.at(i), ocpp::DateTime(period_end_time)};
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: Can you update this with the struct fields:
` {.period = periods.at(i) .end_time = ocpp::DateTime(period_end_time)};`

File: lib/ocpp/v201/smart_charging.cpp [hunk 7]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+                log_period_date_time_pair(date_time_pair);
+                return date_time_pair;
+            }
+            period_start_time.emplace(ocpp::DateTime(period_end_time));
+        }
+    }
+
+    PeriodDateTimePair date_time_pair = {std::nullopt, MAX_DATE_TIME};
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: Same struct fields comment.

File: lib/ocpp/v201/smart_charging.cpp [hunk 8]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+                    current_purpose_and_stack_limits.at(profile.chargingProfilePurpose).stack_level = stackLevel;
+                } else {
+                    // skip profiles with a lower stack level if a higher stack level already has a limit for this
+                    // point in time
+                    continue;
+                }
+            }
+        };
+
+        // break;
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: Can we remove this?

File: lib/ocpp/v201/smart_charging.cpp [hunk 9]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    } else {
+        return MAX_DATE_TIME;
+    }
+}
+
+bool continue_time_arrow(const ocpp::DateTime& temp_time, const ocpp::DateTime& period_start_time,
+                         const ocpp::DateTime& period_end_time, const ocpp::DateTime& lowest_next_time) {
+    // original
+    // return (temp_time >= period_start_time && temp_time < period_end_time && period_end_time < lowest_next_time);
+
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: Does this still need to be there?

File: lib/ocpp/v201/smart_charging.cpp [hunk 10]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    composite_schedule.chargingSchedulePeriod = periods;
+
+    return composite_schedule;
+}
+
+ocpp::DateTime get_current_period_end_time(const ocpp::DateTime& period_start_time,
+                                           std::optional<int32_t> charging_schedule_duration,
+                                           const ChargingSchedulePeriod& period) {
+    return period_start_time;
+}
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: NIT: Can you comment the need for this.

File: lib/ocpp/v201/smart_charging.cpp [hunk 11]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+                                                    // level for every purpose
+        ocpp::DateTime temp_period_end_time;
+        int32_t temp_number_phases;
+
+        for (const ChargingProfile& profile : valid_profiles) {
+            EVLOG_info << "ProfileId #" << profile.id << " Kind: " << profile.chargingProfileKind;
+
+            if (profile.stackLevel > current_purpose_and_stack_limits.at(profile.chargingProfilePurpose).stack_level) {
+                // EVLOG_info << "calculate_composite_schedule> profile.stackLevel > current_purpose_and_stack_limits";
+                EVLOG_info << "boop";
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: NIT: Do we need to boop anymore?

File: lib/ocpp/v201/smart_charging.cpp [hunk 12]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+            current_period_limit = significant_limit_stack_level_pair.limit;
+        }
+        temp_time = this->get_next_temp_time(temp_time, valid_profiles, evse_id);
+    }
+
+    // update duration if end time of last period is smaller than requested end time
+    if (last_period_end_time.to_time_point() - start_time.to_time_point() <
+        (end_time.to_time_point() - start_time.to_time_point())) {
+        composite_schedule.duration =
+            duration_cast<seconds>(last_period_end_time.to_time_point() - start_time.to_time_point()).count();
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: Should this be `determine_duration`

File: lib/ocpp/v201/smart_charging.cpp [hunk 13]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+        return ocpp::DateTime(period_start_time.to_time_point() + seconds(period_diff_in_seconds));
+    } else {
+        return MAX_DATE_TIME;
+    }
+}
+
+bool continue_time_arrow(const ocpp::DateTime& temp_time, const ocpp::DateTime& period_start_time,
+                         const ocpp::DateTime& period_end_time, const ocpp::DateTime& lowest_next_time) {
+    // original
+    // return (temp_time >= period_start_time && temp_time < period_end_time && period_end_time < lowest_next_time);
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp:631]: NIT: Remove commented out code?

File: lib/ocpp/v201/smart_charging.cpp [hunk 14]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+                    EVLOG_info << "period.has_value() stackLevel = " << limit;
+
+                    // update data structure with limit and stack level for this profile
+                    current_purpose_and_stack_limits.at(profile.chargingProfilePurpose).limit = limit;
+                    current_purpose_and_stack_limits.at(profile.chargingProfilePurpose).stack_level = stackLevel;
+                } else {
+                    // skip profiles with a lower stack level if a higher stack level already has a limit for this
+                    // point in time
+                    continue;
+                }
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: This `continue` doesn't seem necessary here - we're at the end of the loop already, so we're already going to continue in either case.

File: lib/ocpp/v201/smart_charging.cpp [hunk 15]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    std::optional<ocpp::DateTime> period_start_time;
+
+    if (startSchedule) {
+        // TODO ?: What if startSchedule is not set? This use case is not handled in the 1.6 code
+        const ocpp::DateTime start_schedule =
+            ocpp::DateTime(std::chrono::floor<seconds>(startSchedule.value().to_time_point()));
+        int seconds_to_go_back;
+
+        if (recurrencyKind.value() == RecurrencyKindEnum::Daily) {
+            seconds_to_go_back = duration_cast<seconds>(time.to_time_point() - start_schedule.to_time_point()).count() %
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: The duration calculation part could be `determine_duration`

File: lib/ocpp/v201/smart_charging.cpp [hunk 16]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+                                 (HOURS_PER_DAY * SECONDS_PER_HOUR);
+        } else {
+            seconds_to_go_back = duration_cast<seconds>(time.to_time_point() - start_schedule.to_time_point()).count() %
+                                 (SECONDS_PER_DAY * DAYS_PER_WEEK);
```
Comment (ghost) [inline lib/ocpp/v201/smart_charging.cpp]: The duration calculation part could be determine_duration

File: lib/ocpp/v201/smart_charging.cpp [hunk 17]
```diff
@@ -402,4 +403,470 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+                if (period_date_time_pair.period.has_value()) {
+                    const ChargingSchedulePeriod period = period_date_time_pair.period.value();
+
+                    temp_period_end_time = period_date_time_pair.end_time;
+
+                    if (period.numberPhases) {
+                        temp_number_phases = period.numberPhases.value();
+                    } else {
+                        temp_number_phases = DEFAULT_AND_MAX_NUMBER_PHASES;
+                    }
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: This could be:
```suggestion
                    temp_number_phases = period.numberPhases.value_or(DEFAULT_AND_MAX_NUMBER_PHASES);
```

File: tests/lib/ocpp/v201/smart_charging_test_utils.hpp [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+#include <sstream>
+#include <vector>
+
+namespace ocpp::v201 {
+
+static const std::string BASE_JSON_PATH = "/tmp/EVerest/libocpp/v201/json/";
+
+class SmartChargingTestUtils {
+public:
+    static std::vector<ChargingProfile> getChargingProfilesFromDirectory(const std::string& path) {
```
Comment (christopher-davis-afs) [inline tests/lib/ocpp/v201/smart_charging_test_utils.hpp]: Function names need to be `snake_case` (Everest style thing)

File: tests/lib/ocpp/v201/test_composite_schedule.cpp [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+    SmartChargingHandler handler = create_smart_charging_handler();
+    boost::uuids::random_generator uuid_generator = boost::uuids::random_generator();
+};
+
+/**
+ * Validates the SmartChargingHandler::determined_duraction and SmartChargingHandler::within_time_window
+ * utility functions.
+ */
+TEST_F(ChargepointTestFixtureV201, K08_CalculateCompositeSchedule_DetermineDurationAndWithinTimeWindow) {
+    GTEST_SKIP();
```
Comment (christopher-davis-afs) [inline tests/lib/ocpp/v201/test_composite_schedule.cpp]: Should these tests have `GTEST_SKIP()`?

File: tests/lib/ocpp/v16/test_composite_schedule.cpp [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+/**
+ * CompositeSchedule Test Fixture
+ */
+class CompositeScheduleTestFixture : public testing::Test {
+protected:
+    void SetUp() override {
+        this->evse_security = std::make_shared<EvseSecurityMock>();
+    }
+
+    void addConnector(int id) {
```
Comment (ghost) [inline tests/lib/ocpp/v16/test_composite_schedule.cpp]: The community prefers snake_case for function names. I am not sure that we are consistent with this in our test files thought. If not you can ignore this comment.

File: tests/lib/ocpp/v201/json/TXProfile_Relative_1.json
Comment (ghost) [inline tests/lib/ocpp/v201/json/TXProfile_Relative_1.json:1]: NIT: Is this file needed?

File: tests/lib/ocpp/v201/test_composite_schedule.cpp [hunk 21]
```diff
@@ -0,0 +1,431 @@
...
+        return ChargingSchedule{
+            id,
+            charging_rate_unit,
+            charging_schedule_period,
+            custom_data,
+            start_schedule,
+            duration,
+            min_charging_rate,
+            sales_tariff,
+        };
```
Comment (ghost) [inline tests/lib/ocpp/v201/test_composite_schedule.cpp]: NIT: Can you update this with the struct fields

**Review by ghost — CHANGES_REQUESTED** [review]
Lets mob on the comments.

File: include/ocpp/v201/smart_charging.hpp (hunk 1 above)
Comment (folkengine) [reply include/ocpp/v201/smart_charging.hpp]: 'Tis done.

File: lib/ocpp/v201/smart_charging.cpp (hunk 13 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp:631]: Don't want to remove this yet. Added `// TODO: Add test coverage to ensure that these calculations are correct before removing this comment.`

File: lib/ocpp/v201/smart_charging.cpp (hunk 11 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp]: Yes, yes we do.

File: lib/ocpp/v201/smart_charging.cpp (hunk 10 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp]: No. There is none.

File: tests/lib/ocpp/v201/json/TXProfile_Relative_1.json
Comment (folkengine) [reply tests/lib/ocpp/v201/json/TXProfile_Relative_1.json:1]: Nope

File: tests/lib/ocpp/v201/test_composite_schedule.cpp (hunk 19 above)
Comment (folkengine) [reply tests/lib/ocpp/v201/test_composite_schedule.cpp]: Removed all but one that will need to triage.

**Review by ghost — APPROVED** [review]
LGTM - ☕

File: lib/ocpp/v201/smart_charging.cpp [hunk 22]
```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    log_period_date_time_pair(date_time_pair);
+    return date_time_pair;
+}
+
+///
+/// \brief Calculates the composite schedule for the given \p valid_profiles and the given \p connector_id
+///
+/// Step 01-000 - Pass in profiles, start and end time, evse od add charging_rate_unit
+/// Step 01-001 - initialize_composite_schedule(start_time, end_time, evse_id, charging_rate_unit)
+/// Step 01-002 - Instantiate vector of ChargingSchedulePeriod periods
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: > evse od add charging_rate_unit

Not sure what this is supposed to say

File: lib/ocpp/v201/smart_charging.cpp [hunk 23]
```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    std::vector<ChargingSchedulePeriod> periods;
+
+    ocpp::DateTime temp_time(start_time);
+    ocpp::DateTime last_period_end_time(end_time);
+    int current_period_limit = MAX_PERIOD_LIMIT;
+    LimitStackLevelPair significant_limit_stack_level_pair = {MAX_PERIOD_LIMIT, -1};
+
+    // calculate every ChargingSchedulePeriod of result within this while loop
+    while (SmartChargingHandler::within_time_window(temp_time, end_time)) {
+        // while (duration_cast<seconds>(end_time.to_time_point() - temp_time.to_time_point()).count() > 0) {
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: Looks like this line is replaced by `within_time_window()`?

File: lib/ocpp/v201/smart_charging.cpp [hunk 24]
```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    return date_time_pair;
+}
+
+///
+/// \brief Calculates the composite schedule for the given \p valid_profiles and the given \p connector_id
+///
+/// Step 01-000 - Pass in profiles, start and end time, evse od add charging_rate_unit
+/// Step 01-001 - initialize_composite_schedule(start_time, end_time, evse_id, charging_rate_unit)
+/// Step 01-002 - Instantiate vector of ChargingSchedulePeriod periods
+CompositeSchedule SmartChargingHandler::calculate_composite_schedule(std::vector<ChargingProfile> valid_profiles,
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp:484]: This is something I mentioned before: why take in `valid_profiles` instead of retrieving them from the `SmartChargingHandler` ourselves? That way we can know all of the profiles we're using to calculate a composite schedule are valid - allowing us to remove some checks that are further down.

File: lib/ocpp/v201/smart_charging.cpp [hunk 25]
```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+                SmartChargingHandler::determine_duration(start_schedule, time) % (HOURS_PER_DAY * SECONDS_PER_HOUR);
+        } else {
+            seconds_to_go_back =
+                SmartChargingHandler::determine_duration(start_schedule, time) % (SECONDS_PER_DAY * DAYS_PER_WEEK);
+        }
+        auto time_minus_seconds_to_go_back = time.to_time_point() - seconds(seconds_to_go_back);
+        period_start_time.emplace(ocpp::DateTime(time_minus_seconds_to_go_back));
+    } else {
+        EVLOG_warning << "Recurring profile with no startSchedule, this should not be possible";
+    }
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp:750]: So, the only case where we wouldn't have a `startSchedule` is if something has gone seriously wrong. As mentioned above, if we know the profile is valid we don't need to double down on checking for/handling invalid cases.

File: lib/ocpp/v201/smart_charging.cpp [hunk 26]
```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    return period_start_time;
+}
+
+std::optional<ocpp::DateTime> SmartChargingHandler::get_relative_profile_start_time(const int32_t evse_id) {
+    std::optional<ocpp::DateTime> period_start_time;
+
+    if (this->evses.at(evse_id)) {
+        EVLOG_debug << "And we are in!";
+    }
+
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: Is this debug output still required? There's not much context here and it feels odd.

File: lib/ocpp/v201/smart_charging.cpp [hunk 27]
```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+    return period_start_time;
+}
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: `period_start_time` as written will always be blank - why do we have a function just to create and return a blank value?

File: lib/ocpp/v201/smart_charging.cpp [hunk 28]
```diff
@@ -402,4 +403,436 @@ void SmartChargingHandler::conform_validity_periods(ChargingProfile& profile) co
...
+
+            //                                             if profile_receipt_time <=
+            //                                             transaction_profile_activation_time
+            //     : return transaction_profile_activation_time else : return profile_receipt_time
+
+            // TODO
+            // if (this->evses.at(evse_id)->has_active_transaction()) {
+            //     period_start_time.emplace(ocpp::DateTime(floor<seconds>(
+            //         this->evses.at(evse_id)->get_transaction()->get_start_energy_wh()->timestamp.to_time_point())));
+            // }
```
Comment (christopher-davis-afs) [inline lib/ocpp/v201/smart_charging.cpp]: Should we create a separate card for this?

File: tests/lib/ocpp/v201/CMakeLists.txt [hunk 29]
```diff
@@ -7,5 +7,10 @@ target_sources(libocpp_unit_tests PRIVATE
...
         test_ocsp_updater.cpp
         test_component_state_manager.cpp
+        test_composite_schedule.cpp
         test_device_model.cpp
-        test_smart_charging_handler.cpp)
+        test_smart_charging_handler.cpp
+        smart_charging_test_utils.hpp
+        )
+
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/json DESTINATION /tmp/EVerest/libocpp/v201)
```
Comment (christopher-davis-afs) [inline tests/lib/ocpp/v201/CMakeLists.txt:16]: We probably want to install this somewhere in the build directory or the install prefix instead of hardcoding to `/tmp`.

File: tests/lib/ocpp/v201/test_composite_schedule.cpp [hunk 30]
```diff
@@ -0,0 +1,441 @@
...
```
Comment (christopher-davis-afs) [inline tests/lib/ocpp/v201/test_composite_schedule.cpp]: NIT: Could use the struct field syntax we use elsewhere

**Review by christopher-davis-afs — CHANGES_REQUESTED** [review]

**Comment by christopher-davis-afs** [comment]
For this PR can you squash down the commits? That way the feature branch gets one clean commit for this - other developers don't need to see the commits related to our review comments.

File: lib/ocpp/v201/smart_charging.cpp (hunk 24 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp:484]: We can do this restructuring, but it will cause a significant delay to this merge. I would suggest making this a separate card, since the goal was to get this code in, and not devote a lot of time to restructuring. 

Personally, I would rather take the opposite approach and decouple the composite schedule logic altogether from the handler.

File: lib/ocpp/v201/smart_charging.cpp (hunk 25 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp:750]: I prefer to avoid assuming that other parts of systems always work as expected. I would prefer to reject this recommendation, and discuss refactorings at a later date.

File: lib/ocpp/v201/smart_charging.cpp (hunk 27 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp]: This is for later work regarding relative profiles.

File: tests/lib/ocpp/v201/CMakeLists.txt (hunk 29 above)
Comment (folkengine) [reply tests/lib/ocpp/v201/CMakeLists.txt:16]: As discussed before, I have not found a good method to importing relative to the project. This is a great refactoring point for after the merge.

File: lib/ocpp/v201/smart_charging.cpp (hunk 28 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp]: yes.

File: lib/ocpp/v201/smart_charging.cpp (hunk 26 above)
Comment (folkengine) [reply lib/ocpp/v201/smart_charging.cpp]: This is for the future relative profile work.

**Review by christopher-davis-afs — APPROVED** [review]
